samples/DirectX/D3D12/HelloConstBuffer12.cs

```diff
@@ -0,0 +1,807 @@
+        private readonly ID3D12Resource*[] _renderTargets;
+        private ID3D12CommandAllocator* _commandAllocator;
+        private ID3D12CommandQueue* _commandQueue;
+        private ID3D12RootSignature* _rootSignature;
+        private ID3D12DescriptorHeap* _rtvHeap;
+        private ID3D12DescriptorHeap* _cbvHeap;
+        private ID3D12PipelineState* _pipelineState;
+        private ID3D12GraphicsCommandList* _commandList;
+        private uint _rtvDescriptorSize;
+        private uint _cbvDescriptorSize;
+
+        // App resources.
+        private ID3D12Resource* _vertexBuffer;
+        private D3D12_VERTEX_BUFFER_VIEW _vertexBufferView;
+        private ID3D12Resource* _constantBuffer;
+        private SceneConstantBuffer _constantBufferData;
+        private byte* _pCbvDataBegin;
+
+        // Synchronization objects.
+        private uint _frameIndex;
+        private HANDLE _fenceEvent;
+        private ID3D12Fence* _fence;
+        private ulong _fenceValue;
+
+        public HelloConstBuffer12(uint width, uint height, string name)
+            : base(width, height, name)
+        {
+            _renderTargets = new ID3D12Resource*[FrameCount];
+
+            _viewport = new D3D12_VIEWPORT {
+                TopLeftX = 0,
+                TopLeftY = 0,
+                Width = width,
+                Height = height,
+                MinDepth = D3D12_MIN_DEPTH,
+                MaxDepth = D3D12_MAX_DEPTH
+            };
+
+            _scissorRect = new RECT {
+                left = 0,
+                top = 0,
+                right = unchecked((int)width),
+                bottom = unchecked((int)height)
+            };
+            _constantBufferData = new SceneConstantBuffer();
+        }
+
+        public override void OnInit()
+        {
+            LoadPipeline();
+            LoadAssets();
+        }
+
+        // Update frame-based values.
+        public override void OnUpdate()
+        {
+            const float translationSpeed = 0.005f;
+            const float offsetBounds = 1.25f;
+
+            _constantBufferData.offset.X += translationSpeed;
+            if (_constantBufferData.offset.X > offsetBounds)
+            {
+                _constantBufferData.offset.X = -offsetBounds;
+            }
+            var readRange = new D3D12_RANGE(); // We do not intend to read from this resource on the CPU.
+            fixed (byte** ppCbvDataBegin = &_pCbvDataBegin)
+            {
+                fixed (SceneConstantBuffer * pConstantBufferData = &_constantBufferData) {
+                    ThrowIfFailed(nameof(ID3D12Resource._Map), _constantBuffer->Map(Subresource: 0, &readRange, (void**)ppCbvDataBegin));
+                    Unsafe.CopyBlock((void*)*ppCbvDataBegin, (void*)pConstantBufferData, (uint)sizeof(SceneConstantBuffer)); // C++: memcpy(_pCbvDataBegin, &_constantBufferData, sizeof(_constantBufferData));
+                    _constantBuffer->Unmap(0, null);
+                }
+            }
```

Comment:
Native code isn't mapping/unmapping and this can all be simplified to just:
```suggestion
            Unsafe.CopyBlock(ref _pCbvDataBegin[0], ref Unsafe.As<SceneConstantBuffer, byte>(ref _constantBufferData), (uint)sizeof(SceneConstantBuffer));
```